orte/mca/odls/singularity/odls_singularity_component.c

```diff
@@ -0,0 +1,117 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2004-2008 The Trustees of Indiana University and Indiana
+ *                         University Research and Technology
+ *                         Corporation.  All rights reserved.
+ * Copyright (c) 2004-2006 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
+ *                         University of Stuttgart.  All rights reserved.
+ * Copyright (c) 2004-2005 The Regents of the University of California.
+ *                         All rights reserved.
+ * Copyright (c) 2015      Los Alamos National Security, LLC. All rights
+ *                         reserved.
+ * Copyright (c) 2017      Intel, Inc. All rights reserved.
+ * Copyright (c) 2019      Sylabs, Inc. All rights reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ *
+ * These symbols are in a file by themselves to provide nice linker
+ * semantics.  Since linkers generally pull in symbols by object
+ * files, keeping these symbols as the only symbols in this file
+ * prevents utility programs such as "ompi_info" from having to import
+ * entire components just to query their version and parameters.
+ */
+
+#include "orte_config.h"
+#include "orte/constants.h"
+
+#include <stdlib.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+#include <ctype.h>
+
+#include "orte/mca/mca.h"
+#include "opal/mca/base/base.h"
+
+#include "orte/mca/odls/odls.h"
+#include "orte/mca/odls/base/odls_private.h"
+#include "orte/mca/odls/singularity/odls_singularity.h"
+
+int odls_singularity_priority = 9;
```

Comment:
IIRC, non static global variables are prefixed with `mca_` or possibly `orte_`.

Reply:
True - I will fix.

Reply:
> Yeah, I've been working thru this ODLS component and comparing it to the old schizo component from OMPI v3.1. I think this is better implemented by modifying the old schizo component - it appears to require only a small update. The schizo framework is already setup as multi-select and so it allows for PRRTE to support both Singularity and non-Singularity applications at the same time.

I have to admit that I do not know what you are referring to, but I trust you. :)

> 
> I can bring that code over to PRRTE and remove this component, and I can update the schizo component with what appears to be the new requirements. However, it does still beg the question of "how do we know a Singularity container is being launched?". It appears that we can use the presence of `SY_EXEC_ARGS` in the environment as an indicator, or perhaps see if the container name ends in `.si` as @jjhursey suggests.
> 
> Any thoughts?

The goal of this PR (and I should have included it in the description, sorry) is to provide all the nobs to let users and a few tools to start Singularity containers. The main challenges are
- based on what is in the image and how the image is build, the flags to start the containers may change. For instance, if the application is writing to the FS, one should use `--write` as a `singularity exec`
- I have a set of tools that make it easier to install Singularity, as well as create, manage and execute containers. Based on how the containers end up being create, different `exec` flags are required. For example, the default for one of my tools is `--no-home --contain --cleanenv --writable`.
This is what `SY_EXEC_ARGS` is about: give the opportunity to the user/tool that is running `prun` to pass down flags/arguments to the `singularity exec` command.